### PR TITLE
Implement option to gather only local filestore stats.

### PR DESCRIPTION
**Pull request by Space2Man**
Initial version. Please check and verify ... did not even test it yet ...

### CODE REVIEW

File: oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java [hunk 1]
```diff
@@ -163,6 +175,10 @@ private boolean listElementStartsWith(List<String> aList, String charSeq) {
                 continue;
             }
 
+            if (localOnly && (type.startsWith("nfs") || type.equals("cifs"))) {
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java]: Probably should add `smbfs` and `sshfs` for completeness (here and other places these are used).  Actually found this list...

`afs, cifs, smbfs, sshfs, ncpfs, ncp, nfs, nfs4, gfs, gds2, glusterfs`

Plus `afp` on mac, although looks like we handle that separately.

Might want a protected static `isNetworkFsType(String type)` method in the abstract parent.

File: oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java [hunk 2]
```diff
@@ -107,6 +106,9 @@
                     if (volume.equals("devfs") || volume.startsWith("map ")) {
                         continue;
                     }
+                    if (localOnly && (volume.startsWith("localhost:") || volume.startsWith("//"))) {
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java]: I'll do some testing later but I'm iffy about this filter.  We do have a `LOCAL_DISK` pattern matcher for the `/dev` path that might be appropriate here.

File: oshi-core/src/main/java/oshi/software/os/unix/solaris/SolarisFileSystem.java [hunk 3]
```diff
@@ -105,7 +104,11 @@ private boolean listElementStartsWith(List<String> aList, String charSeq) {
         String key = null;
         String total = null;
         String free = null;
-        for (String line : ExecutingCommand.runNative("df -g")) {
+        String command = "df -g";
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/unix/solaris/SolarisFileSystem.java]: These 4 lines might read more simply with a ternary operator.

File: oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java (hunk 2 above)
Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java]: Yep, this is incomplete.  There's at least `smb://192.168.foo.bar/disk` for example and a `foobar@oxfuse`.   And the `/dev` search might not screen out some non-local systems.

Some options are to parse `df -l` similar to the *nixes, parse `mount` looking for the 'local' flag, or use the `f_fstypename` field to grab the string for the same test as the other *nix systems.

I'll dig into the source code for `mount` or `df` later to see if I can figure out what it uses internally.

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java]: Aha, we have an `f_flags` field already, which we can bitmask with `MNT_LOCAL` (`0x00001000`).  Which makes me realize that perhaps we should reopen #273.

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java]: Here's the actual code implementing the `l` flag in `df -l`

```
if (lflag && (mntbuf[i].f_flags & MNT_LOCAL) == 0)
			continue;
```

File: oshi-core/src/main/java/oshi/software/os/windows/WindowsFileSystem.java [hunk 4]
```diff
@@ -141,7 +136,9 @@ public WindowsFileSystem() {
                 volumeMap.get(wmiVolume.getMount()).setName(wmiVolume.getName());
             } else {
                 // Otherwise add the new volume in its entirety
-                result.add(wmiVolume);
+                if (!localOnly) {
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/software/os/windows/WindowsFileSystem.java]: This can be moved up as an `else if`.

**Comment by dbwiddis** [comment]
I went ahead and added the macOS code and more thorough fs type checking for the other *nix systems.  Still TODO is [filter the WMI query to only ask for the local drives](https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-logicaldisk).   We could query on the `DriveType`: (Unknown (0), No Root Directory (1), Removable Disk (2), Local Disk (3), Network Drive (4), Compact Disc (5), RAM Disk (6)) but should decide whether to include only 3, or just exclude 4 (my vote).  The `MediaType` field is another option (non-zero?) but I think isn't as good as `DriveType`.

**Comment by dbwiddis** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- oshi-core/src/main/java/oshi/software/os/unix/solaris/SolarisFileSystem.java  4
- oshi-core/src/main/java/oshi/software/common/AbstractFileSystem.java  3
- oshi-core/src/main/java/oshi/software/os/unix/freebsd/FreeBsdFileSystem.java  2
- oshi-core/src/main/java/oshi/software/os/mac/MacFileSystem.java  3
- oshi-core/src/main/java/oshi/software/os/windows/WindowsFileSystem.java  3
- oshi-core/src/main/java/oshi/software/os/linux/LinuxFileSystem.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/widdis/oshi/pullRequest?prid=4976929&bid=16439462)

**Comment by dbwiddis** [comment]
I think I'm done tweaking your PR.  I'll merge tonight unless you have anything else you'd like to comment on it.

**Comment by Space2Man** [comment]
Man, you are fast! I hope to be able to test it on one of our systems in the next few days.

Thanks a lot!

**Pull request merged**